### ISSUE TITLE
Support for newterm

### ISSUE THREAD

**Issue by gizmomogwai** [mpevnev/nice-curses#7]
Hi,

I was looking into keeping stdin open for some input into a program, but at the same time I already want to show and use an ncurses ui.
I understand that with ncurses directly I could use newterm. Could you please package that nicely into nice-curses as well?

**Comment by mpevnev** [comment]
Oh wow, that's one hell of an oversight on my part. This might take a while, and will probably be a breaking change.

**Comment by gizmomogwai** [comment]
Breaking change is not a problem from me. I am not very familiar with ncurses and was researching in this direction, but if you could add this, it would be great.
i also could try to get rid of you nice wrapper around the deimos ncurses thing (but actually i like you api).

**Comment by gizmomogwai** [comment]
i think its might not be soo bad.
it looks like only `stdscr = new Window(null, initscr());` should get a little bit more generic?

**Comment by mpevnev** [comment]
That'll be a part of it, yes. There will also have to be a `Screen` class, an array of those in `Curses` for clean-up and a separate constructor for `Curses` which doesn't invoke `initscr`. But that's kind of a minor thing. The main part will be to do some research and experimenting on how multiple screens work in ncurses, so the surprises and inconveniences can be minimized. I'm worried in particular about relationship between windows and screens - whether a window is always drawn on the current screen, or whether it is somehow associated with the screen that was current when it was created, and will only affect that particular screen; this is my main concern. (btw, if you happen to know this, please do share)

**Comment by gizmomogwai** [comment]
Unfortunately I am not experienced enough with ncurses ... But perhaps I can learn.

**Comment by mpevnev** [comment]
Started experimenting. It seems that windows do get associated with the terminal that was active when they were created: [this program](https://gist.github.com/mpevnev/f82bace58b8f8b41afb9510b02757bb6) outputs only "foobar" or "baz", depending on which terminal was `set_term`ed before the call to `doupdate` (note that `newterm` sets the freshly created terminal as the active one).

Also, the manpage for `endwin` says that if several terminals use the same file, they should be disposed of in the reverse order of their creation. This is nasty, because it effectively forbids deleting any screen except the most recent one unless I'm 100% sure its underlying files are not used anywhere else. (Indeed, if I switch the order of clean-up in the program in the linked gist, things go horribly wrong - the terminal is not restored to normal settings after the program is done, and moreover I dread some UB is involved). I hope that `fileno` thing on `std.stdio.File` is both cross-platform and does uniquely identify the file in question (so I can call the actual ncurses functions only if it's safe), as it would allow to make screen-removing functions public without handing the user such a nice way to shoot ximself in the foot. Otherwise keeping them private is probably the best bet, or maybe having a specialized `Curses` subclass that would allow it (for adventurous souls).

**Comment by gizmomogwai** [comment]
Thanks for looking into that.
My main usecase at the moment would not include several screens and windows and what not.
I just want to keep reading from stdin, but already have a ncurses ui open on the data that just arrived.

**Comment by mpevnev** [comment]
I'm sorry, I can't do this. A month ago I hit something suspiciously similar to classic undefined behaviour of C. I ran the test suite through `valgrind`, and it complained about uninitialized variables being used. In an attempt to fix it, I tried to completely revise the resource management of the library. To do so, I had to dig deeper into D, and I don't like it at all.

A few modules in the standard library are not documented at all. A few are not even visible at the sidebar of the Phobos page. `core.lifetime` is actually _missing_ from the version of stdlib that comes with `ldc`. And the worst of all, the 'uninitialized values' warning, as I've leared today, wasn't caused by my abuse of `ncurses`! Just creating a garbage-collected array is enough to trigger it.

In short, I'm not going to touch D with a ten-foot pole from now on.

**Comment by gizmomogwai** [comment]
@mpevnev thanks for giving it a shot anyways.
I also do have some issues with dlang (although they are not as severe as yours).
Would be nice to run dlang stuff with valgrind, but I am not sure that even a simple hello world program would work there.
